### ISSUE TITLE
ManifestManager.get_depends & get_depends_on should not depend on rosdep

### ISSUE THREAD

**Issue by abencz** [ros-infrastructure/rospkg#103]
Both of these queries can be implemented by examining the the list of packages in the ROS environment and filtering out non-ROS dependencies in the manifest based on this list. This is important because these methods are used in several places at runtime (e.g. to load plugins in roswtf) and runtime behaviour shouldn't depend on rosdeps being up-to-date.

**Comment by dirk-thomas** [comment]
How exactly should the question "filtering out non-ROS dependencies" be answered without rosdep in this case?

**Comment by abencz** [comment]
Actually, you're right, this is probably not possible for ManifestManager.get_depends, but get_depends_on relies on looking through the manifests of packages on the system. It shouldn't need to check whether the dependencies in those manifests are ROS dependencies since it's looking for a specific dependency name already.

**Comment by abencz** [comment]
Updated title.

**Comment by dirk-thomas** [comment]
That sounds reasonable. Since you seem to have a use case for this could you please consider providing a pull request for this enhancement.

**Comment by abencz** [comment]
Ok, will attempt do to so in the next few day.

**Comment by mikepurvis** [comment]
This was originally motivated by a plugin discovery use-case, but we have another one now— rosbag_migration uses get_depends_on, and therefore depends on rosdep being present and initialized: https://github.com/ros/ros_comm/blob/kinetic-devel/tools/rosbag/src/rosbag/migration.py#L529

FYI @efernandez

**Comment by efernandez** [comment]
If I understand well, we need to remove the dependency on `rosdep2` re-implementing this without using it, right?
https://github.com/ros-infrastructure/rospkg/blob/master/src/rospkg/manifest.py#L389

I don't fully understand which other tools/libraries can be used instead. If any of you could help me with that, I could try to provide a PR.

**Comment by efernandez** [comment]
Essentially, these are the only two methods used from `rosdep2` that we need to replace:
https://github.com/ros-infrastructure/rosdep/blob/master/src/rosdep2/rospack.py#L66
https://github.com/ros-infrastructure/rosdep/blob/master/src/rosdep2/rospack.py#L70

So the question is: how can we check if a dependency is a ros OR a system dependency w/o `rosdep2`?

@abencz Could you elaborate a little more on why we don't need to filter ros from system dependencies for `get_depends_on`?

**Comment by efernandez** [comment]
ping @abencz  @dirk-thomas

**Comment by mikepurvis** [comment]
The proposal made above is that `get_depends_on` should be reimplemented so that rather than using `get_depends` under the hood (which requires checking what type of dep each one is), it can instead simply return a list of all known packages expressing a direct dependency on the given key.

This should be pretty easy— it's essentially popping open each `package.xml` (or `manifest.xml`) file pointed to by `rospack list` and checking whether it contains the specified dependency.

The main questions I'd have for @dirk-thomas about this is:

- Is there functionality in `catkin_pkg` that we should be leveraging for this?
- To maintain compatibility for rosbuild packages, would we want to do something like maintain the current/legacy implementation, but only call it when a rosbuild package is detected in the workspace? Is there some better way to handle this?

**Comment by dirk-thomas** [comment]
> Is there functionality in `catkin_pkg` that we should be leveraging for this, other than just the `package.xml` parsing?

No, `catkin_pkg` only provides the parsing and some abstract for various dependency types.

> To maintain compatibility for rosbuild packages, would we want to do something like maintain the current/legacy implementation, but only call it when a rosbuild package is detected in the workspace? Is there some better way to handle this?

I don't think a patch should call `catkin_pkg` directly. The existing method `ManifestManager.get_manifest` should be sufficient to use and already abstracts between `catkin` and `rosbuild` manifests.

Currently `get_depends` is being called with `implicit=True` which performs the recursion from top to bottom (which requires `rosdep` to distinguish the type of the rosdep key). `get_depends_on` needs to reimplement that recursion but from bottom to top - then `rosdep` isn't required since the "step upwards" in the dependency hierarchy is always from a ROS package to another ROS package.

**Comment by mikepurvis** [comment]
I've now gone down the rabbit hole a bit on this, and it's actually a little different than expressed above.

Both rosbag migration plugin discovery and roswtf plugin discovery [call `get_depends_on` with `implicit=False`](https://github.com/ros/ros_comm/search?q=get_depends_on), so based on the implementation that's there, there's no reason for `get_depends` to ever get called under those circumstances— and changing the implicit implementation would have no effect:

```
    def get_depends_on(self, name, implicit=True):
        depends_on = []
        if not implicit:
            for r in self.list():
                if r == name:
                    continue
                try:
                    m = self.get_manifest(r)
                    if any(d for d in m.depends if d.name == name):
                        depends_on.append(r)
                except [snip]
        else:
            for r in self.list():
              # get recursive dependencies; uses depends_on with the rosdep2 dependency
        return depends_on
```

The _real issue_ is that the sorting out of rosdeps and system package dependencies happens in the `parse_manifest_file` function:

https://github.com/ros-infrastructure/rospkg/blob/60109b662b1cfd0e3fb8d59d1a72069cbbb631b9/src/rospkg/manifest.py#L387-L410

It's not at all clear to me that there's a good way forward here:

- Bypassing get_manifest is no good because then we're skipping over the caching it provides.
- Passing through an argument to skip the dep sorting is no good because then the cache will be inconsistent.
- Silencing the rosdep warning isn't ideal because there are legitimate circumstances where a user should see that warning.
- Adding an env var or some other way to explicit skip this behaviour feels super clunky.

Maybe the rosdeps/depends attributes could be populated lazily rather than upfront, and the check could become `if any(d for d in (m.buildtool_depends + m.build_depends + m.run_depends + m.test_depends) if d.name == name)` or something? This would work, but will require some `__getattr__` trickery which I'm not sure will be acceptable to @dirk-thomas.
